First commit mgmt-app

fgaribay-nearsoft opened the pull request:
This is the first iteration of the Management App. It is a minimal implementation and I would appreciate your input, especially on what you would recommend for me to test.

fgaribay-nearsoft committed:
First commit mgmt-app

carlos-hernandez commented on management-app/.gitignore:
```diff
@@ -0,0 +1,25 @@
+/target/
```
This settings can be included in the global `.gitignore` file so each Maven submodule can reuse them.

carlos-hernandez commented on management-app/.mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1 @@
+distributionUrl=https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip
```
Once you add more Maven submodules to your project you can explore the option to have this wrapper at a project level instead of having it in the submodules.

fgaribay-nearsoft replied:
Thanks, I will include this in the PR where I integrate both maven submodules so I can test it actually working.

carlos-hernandez commented on management-app/pom.xml:
```diff
@@ -0,0 +1,72 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.nearsoft.fgaribay</groupId>
```
Group id in this place is not needed since it's inherited from the parent POM

carlos-hernandez commented on pom.xml:
```diff
@@ -0,0 +1,27 @@
+    <name>skt-task</name>
+    <description>SKT Task Project</description>
+
+    <parent>
```
As a personal preference, I avoid using this Spring Boot parent POM and use the `<dependencyManagement>` tag instead:
-  [Spring Boot Dependency Management with a Custom Parent](https://www.baeldung.com/spring-boot-dependency-management-custom-parent)

Using this approach, make sure that the `spring-boot-maven-plugin` in `management-app/pom.xml` file specifies the plugin version (`1.5.7.RELEASE` instead of `2.0.3.RELEASE`), which you can define as a property in this POM file and reuse it in the submodule.

carlos-hernandez commented on management-app/pom.xml:
```diff
@@ -0,0 +1,72 @@
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <properties>
```
This properties can be moved to the parent POM. Also, it's a good practice to specify the compiler source and target version, too:
- [Setting the -source and -target of the Java Compiler](https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-source-and-target.html)

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/ManagementApplication.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+
+@SpringBootApplication
+public class ManagementApplication extends SpringBootServletInitializer {
```
`SpringBootServletInitializer` is a feature added in Spring Boot 2.0, which is not the version required for this project.

carlos-hernandez commented on management-app/src/main/resources/application.yml:
```diff
@@ -0,0 +1,18 @@
+spring.mvc.view.prefix: /WEB-INF/jsp/
```
Use YAML syntax, like this:

```yaml
spring:
  mvc:
    view:
      prefix: /WEB-INF/jsp/
      suffix: .jsp
```

carlos-hernandez commented on management-app/src/main/resources/static/WEB-INF/jsp/createProduct.jsp:
```diff
@@ -0,0 +1,37 @@
+<!DOCTYPE HTML>
```
Why do you have one version of this JSP in `management-app/src/main/resources/static/WEB-INF/jsp` and another one in `management-app/src/main/webapp/WEB-INF/jsp`?

carlos-hernandez commented on management-app/src/main/resources/static/css/main.css:
```diff
@@ -0,0 +1,7 @@
+h1 {
```
`main.css` and `style.css` files have the same contents. Consider using only one of them and delete the other.

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/model/ProductResponse.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.UUID;
+
+public class ProductResponse {
+    String operation;
```
Mark the fields in this POJO as `private`.

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/BrokerConnector.java:
```diff
@@ -0,0 +1,54 @@
+    @Value("product.list.queue.name")
+    private String productListQueueName;
+
+    @Autowired
```
Since you're using constructor injection the `@Autowired` annotation is not needed. In general, we prefer to use constructor injection instead of field injection, so check your Spring beans and make changes according to this.

carlos-hernandez commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/config/BrokerConfig.java:
```diff
@@ -0,0 +1,26 @@
+
+@Configuration
+public class BrokerConfig {
+    @Value("product.create.queue.name")
```
Explore the usage of `@ConfigurationProperties` to get this values from your settings.
- [Type-safe Configuration Properties](https://docs.spring.io/spring-boot/docs/1.5.7.RELEASE/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties)

carlos-hernandez reviewed (changes requested):
Please check the individual comments and feel free to ping me if you need to clarify some of them.

fgaribay-nearsoft committed:
Refactor .gitignore into parent

fgaribay-nearsoft committed:
Refactor POM into parent

fgaribay-nearsoft committed:
Use the right Spring Boot version

Version 1.5.7-RELEASE is the one used by the actual project.

fgaribay-nearsoft committed:
Remove unnecessary files

fgaribay-nearsoft committed:
Use actual yml syntax

jass-holguin commented on management-app/src/main/webapp/WEB-INF/jsp/index.jsp:
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE HTML>
```
this file is also duplicated in /src/main/resources/static/WEB-INF/jsp

jass-holguin commented on management-app/src/main/webapp/WEB-INF/jsp/listProducts.jsp:
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE HTML>
```
this one is also duplicated in /src/main/resources/static/WEB-INF/jsp

jass-holguin commented on management-app/src/main/java/com/nearsoft/fgaribay/mgmt/ManagementController.java:
```diff
@@ -0,0 +1,40 @@
+
+    @RequestMapping(value = "/create", method = RequestMethod.POST)
+    public String submit(@Valid @ModelAttribute("product") Product product) {
+        brokerConnector.createProduct(product);
```
remember to validate the product, you should probably move the brokerConnector to a service layer, where you would make a validation before sending it to the queue, otherwise if you're planning to do the validation in the microservice, how are you gonna return the error to the mgmt app?

fgaribay-nearsoft replied:
Thanks, you're right, the app should make proper validations. Do you think this is what I should be looking into for implementing this? (https://docs.spring.io/spring/docs/4.1.x/spring-framework-reference/html/validation.html)

I'm thinking of adding a "price" field to the product and have that require some validation logic. As for the strings, I plan on checking for nulls, lengths and only allow alphanumerics and punctuation marks (instead of escaping, to keep the data layer in a black box). Any other suggestions?

Would you prefer to see these changes in a new PR or on this same one?

jass-holguin replied:
You can use that, there's also a Hibernate validator which uses annotations, but you could also just validate with simple IFs directly checking the object, it's not an issue for me, leave a more complex validation if you want for last, if you have time. And yes, you can put it in a new PR :)

fgaribay-nearsoft committed:
Use type-safe configuration properties

fgaribay-nearsoft committed:
Use best practices

- Mark privates where possible
- Autowire on constructor
- Remove unused fields
